mask.cue

```diff
@@ -0,0 +1,20 @@
+package structural
+
+maskrec: { orig: _, mask: _, result: (Mask & { Orig: orig, Mask: mask }).Result }
+Mask :: {
+  Orig: _
+  Mask: _
+  Result: {
+    for k,_ in Orig {
+      if (Builtin & Orig[k]) != _|_ && (Mask[k] & Orig[k]) == _|_ {
+        "\(k)": Orig[k]
+      }
+      if (List & Orig[k]) != _|_ {
+        "\(k)": [v for i, v in Orig[k] if (v & Mask[k][i]) == _|_]
```

Comment:
do we remove the if statement here and call maskrec?

Reply:
we are saying something like "we assume the elements of the list are of the same type, and we want to apply the mask to each element"

There is that option, and the other where we want to mask the list for elements which have all of the fields (unify with the list elem mask).

Is there a way to detect which by comparing the mask against something? Or do we need to "functions" with different semantics?

Reply:
I'm thinking about ways we can generate a large number of test cases, then go through by hand and figure out the answer. Maybe something combinatoric? Maybe some cue playground?

Could be specific to this repo. Hopefully we can use the same data set for all of the different functions. Actually, maybe using some of the paired or tripled calls that result in the identity could be useful... hmm...

Reply:
"Is there a way to detect which by comparing the mask against something? Or do we need to "functions" with different semantics?"

If Orig is a list and Mask isn't, we could unify each element of Orig with the Mask. Then of course if Mask is a list it could go element-wise like it is now?